machine/src/lib.rs

```diff
@@ -12,81 +12,173 @@ macro_rules! errno {
 }
 
 pub mod iface;
-pub mod namespace;
+mod namespace;
+
+pub use namespace::{unshare_user, Namespace};
 
 use async_process::Command;
-use futures::channel::mpsc;
+use futures::channel::{mpsc, oneshot};
+use futures::future::FutureExt;
 use futures::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
 use futures::sink::SinkExt;
 use futures::stream::StreamExt;
 use netsim_embed_core::{Ipv4Range, Packet, Plug};
 use std::fmt::Display;
-use std::io::Write;
+use std::io::{Error, ErrorKind, Result, Write};
 use std::net::Ipv4Addr;
 use std::process::Stdio;
 use std::str::FromStr;
 use std::thread;
 
-#[derive(Clone, Copy, Debug, Eq, PartialEq)]
-pub enum IfaceCtrl {
+#[derive(Debug)]
+enum IfaceCtrl {
     Up,
     Down,
     SetAddr(Ipv4Addr, u8),
 }
 
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct MachineId(pub usize);
+
 /// Spawns a thread in a new network namespace and configures a TUN interface that sends and
 /// receives IP packets from the tx/rx channels and runs some UDP/TCP networking code in task.
-pub fn machine<C, E>(
+#[derive(Debug)]
+pub struct Machine<C, E> {
+    id: MachineId,
     addr: Ipv4Addr,
     mask: u8,
+    ns: Namespace,
+    ctrl: mpsc::UnboundedSender<IfaceCtrl>,
+    tx: mpsc::UnboundedSender<C>,
+    rx: mpsc::UnboundedReceiver<E>,
+    join: Option<thread::JoinHandle<Result<()>>>,
+}
+
+impl<C, E> Machine<C, E>
+where
+    C: Display + Send + 'static,
+    E: FromStr + Send + 'static,
+    E::Err: std::fmt::Debug + Display + Send + Sync,
+{
+    pub async fn new(id: MachineId, plug: Plug, cmd: Command) -> Self {
+        let (ctrl_tx, ctrl_rx) = mpsc::unbounded();
+        let (cmd_tx, cmd_rx) = mpsc::unbounded();
+        let (event_tx, event_rx) = mpsc::unbounded();
+        let (ns_tx, ns_rx) = oneshot::channel();
+        let join = machine(id, plug, cmd, ctrl_rx, ns_tx, cmd_rx, event_tx);
+        let ns = ns_rx.await.unwrap();
+        Self {
+            id,
+            addr: Ipv4Addr::UNSPECIFIED,
+            mask: 32,
+            ns,
+            ctrl: ctrl_tx,
+            tx: cmd_tx,
+            rx: event_rx,
+            join: Some(join),
+        }
+    }
+
+    pub fn id(&self) -> MachineId {
+        self.id
+    }
+
+    pub fn addr(&self) -> Ipv4Addr {
+        self.addr
+    }
+
+    pub fn mask(&self) -> u8 {
+        self.mask
+    }
+
+    pub fn set_addr(&mut self, addr: Ipv4Addr, mask: u8) {
+        self.ctrl
+            .unbounded_send(IfaceCtrl::SetAddr(addr, mask))
+            .unwrap();
+        self.addr = addr;
+        self.mask = mask;
+    }
```

Comment:
This is potentially racy, isn't it. Maybe make it async and wait for `IfaceCtrl` to ack?

Reply:
fixed